src/main/ruby/truffleruby/core/io.rb

```diff
@@ -2415,7 +2415,7 @@ def close
     return nil if closed?
 
     begin
-      flush
+      ensure_open #flush
```

Comment:
This feels wrong, I believe it is correct to `flush` in `close`, and I'd think CRuby does the same.
The difference is CRuby probably calls a C function rather than a method call to `flush`.
It happens to work because IO#flush does basically nothing currently, but that's implementation details and an IO subclass might define flush.

Could you check what CRuby does?

Also, what causes the other call to flush in those RSpec specs?

Reply:
This is the code CRuby which flushes on close:
https://github.com/ruby/ruby/blob/203b1fc7e206764f52c99e44af85c651069cd8b9/io.c#L5337-L5359
Indeed it's not a dynamic call to flush, just a call to fptr_finalize_flush.

So I guess the way it's supposed to work is if you make your own IO subclass or IO-like object you should yourself override close and call flush if necessary, like done in https://github.com/rspec/rspec-core/blob/ea8554afd1a2b63677c6593059fa8f2476181deb/lib/rspec/core/formatters/base_text_formatter.rb#L65-L71